core/color.js

```diff
@@ -0,0 +1,723 @@
+
+				if ( !colorChannels ) {
+					this._.isValidColor = false;
+					return;
+				}
+
+				this._.red = colorChannels[ 0 ];
+				this._.green = colorChannels[ 1 ];
+				this._.blue = colorChannels[ 2 ];
+				this._.alpha = colorChannels[ 3 ];
+			},
+
+			/**
+			 * Get hexadecimal color value based on color name.
+			 *
+			 * @private
+			 * @param {String} colorName color name, e.g. `red`.
+			 * @returns {String/null} Hexadecimal color representation or `null` if such named color does not exists.
+			 */
+			matchStringToNamedColor: function( colorName ) {
+				return CKEDITOR.tools.color.namedColors[ colorName.toLowerCase() ] || null;
+			},
+
+			/**
+			 * Extracts RGBA channels from given HEX string.
+			 *
+			 * @private
+			 * @param {String} colorCode HEX color representation.
+			 */
+			extractColorChannelsFromHex: function( colorCode ) {
+				if ( colorCode.match( CKEDITOR.tools.color.hex3CharsRegExp ) ) {
+					colorCode = this._.hex3ToHex6( colorCode );
+				}
+
+				if ( colorCode.match( CKEDITOR.tools.color.hex6CharsRegExp ) || colorCode.match( CKEDITOR.tools.color.hex8CharsRegExp ) ) {
+					var parts = colorCode.split( '' );
+
+					return [
+						hexToValue( parts[ 1 ] + parts[ 2 ] ),
+						hexToValue( parts[ 3 ] + parts[ 4 ] ),
+						hexToValue( parts[ 5 ] + parts [ 6 ] ),
+						parts[ 7 ] && parts[ 8 ] ? hexToValue( parts[ 7 ] + parts[ 8 ] ) : 1
+					];
+				}
+
+				return null;
+			},
+
+			// TODO
+			extractColorChannelsFromRgba: function() {
+				return null;
+			},
+
+			// TODO
+			extractColorChannelsFromHsla: function() {
+				return null;
+			},
+
+			/**
+			 * Convert 3-characters hexadecimal color format to 6-characters one.
+			 *
+			 * @private
+			 * @param {String} hex3ColorCode 3-characters hexadecimal color, e.g. `#F0F`.
+			 * @returns {String} 6-characters hexadecimal color e.g. `#FF00FF`.
+			 */
+			hex3ToHex6: function( hex3ColorCode ) {
+				var parts = hex3ColorCode.split( '' );
+
+				return '#' + parts[ 1 ] + parts[ 1 ] + parts[ 2 ] + parts[ 2 ] + parts[ 3 ] + parts[ 3 ];
+			},
+
+			/**
+			 * Normalize rgb values into 0-255 range.
+			 *
+			 * @private
+			 * @param {Number} red Number of red channel.
+			 * @param {Number} green Number of green channel.
+			 * @param {Number} blue Number of blue channel.
+			 * @returns {Array} Rgb Array of normalized values.
+			 */
+			normalizeRgbValues: function( red, green, blue ) {
+				function normalizer( number ) {
+					if ( isPercentValue( number ) ) {
+						number = convertPercentValueToNumber( number );
+						number = Math.round( 255 * normalizePercentValue( number ) );
+					} else {
+						number = Number( number );
+						number = number > 255 ? 0 :
+							number < 0 ? 255 : number;
+					}
+					return number;
+				}
+				red = normalizer( red );
+				green = normalizer( green );
+				blue = normalizer( blue );
+
+				return [ red, green, blue ];
+			},
```

Comment:
This method seems to be slightly confusing multi-purpose method. On one hand it just normalizes a number, but for some reason it does it for 3 numbers at once. It is clear in the `color` class context why, but at the same time it seems to be breaking separation of concerns :thinking: Additionally, it handles both integers and percentages differently and the normalization range (255) is hardcoded - both doesn't really arise from method name or its description. So here you have:

1. Normalization of percentage values based on given range (0 - 255).
1. Normalization of numbers form outside 0 - 255 range (not sure why greater numbers results in 0 why negative ones in 255 :thinking:).
1. It is done for 3 numbers at once.

So I think this could be split into smaller simpler methods/functions.

Reply:
As we spoke: we get rid of optimistic input value parsing.
So only valid numbers will be here (percent and proper ranged numbers). 

I'll adjust tests and clean up this method.